yapapi/script/__init__.py

```diff
@@ -0,0 +1,172 @@
+import asyncio
+from datetime import timedelta
+from typing import Any, Awaitable, Callable, Dict, Optional, List, Tuple, TYPE_CHECKING
+
+from yapapi.events import CommandExecuted
+from yapapi.script.capture import CaptureContext
+from yapapi.script.command import (
+    BatchCommand,
+    Command,
+    Deploy,
+    DownloadBytes,
+    DownloadFile,
+    DownloadJson,
+    Run,
+    SendBytes,
+    SendFile,
+    SendJson,
+    Start,
+    Terminate,
+)
+from yapapi.storage import DOWNLOAD_BYTES_LIMIT_DEFAULT
+
+if TYPE_CHECKING:
+    from yapapi.ctx import WorkContext
+
+
+class Script:
+    """Represents a series of commands to be executed on a provider node.
+
+    New commands are added to the script either through its `add` method or by calling one of the
+    convenience methods provided (for example: `run` or `send_json`).
+    Adding a new command *does not* result in it being immediately executed. Once ready, a `Script`
+    instance is meant to be yielded from a worker function (work generator pattern).
+    Commands will be run in the order in which they were added to the script.
+
+    If the `WorkContext` instance this `Script` uses has the field `_implicit_init` set to `True`,
+    the first script to be yielded is going to prepend the user's commands with `Deploy` and
+    `Start` commands.
+    """
+
+    timeout: Optional[timedelta] = None
+    """Time after which this script's execution should be forcefully interrupted."""
+
+    wait_for_results: bool = True
+    """Whether this script's execution should block until its results are available."""
+
+    def __init__(self, context: "WorkContext"):
+        self._ctx: "WorkContext" = context
+        self._commands: List[Tuple[Command, asyncio.Future]] = []
+
+    def _evaluate(self) -> List[BatchCommand]:
```

Comment:
As far as I understand, here (and in few other places) we have a private method that is called from outside because we want to separate "public interface from `yapapi` internals POV" and "public interface for the external developer".

I don't like this. I think the other line of separation is better (called from this class/called from anywhere). But if this is an established convention in `yapapi` than I OK. Is it?

Reply:
I'm not sure if I understand the alternative approach here, could you please elaborate?

Reply:
All methods intended to be called from **outside of the Script class (or possible classes that inherit from it)** are public (that is, they don't start with `_`). So we have `evaluate`, `after` etc. We use the documentation to differentiate between "functions developer should use" and "functions developer shouldn't ever touch".

Reply:
Also: a quite popular solution is to use double underscore for the "really class-private methods". I think that's nice, but only if we really consistently do this.

E.g. `Command._make_batch_command` would have two underscores. Or `_ReceiveContent._emit_download_start`. Or `_ReceiveContent` attributes.

Currently we **sometimes** have two underscores, and that's great, but when we have a function with a single underscore we never know if it is supposed to be called from somewhere else in `yapapi`, or not.

Reply:
There's no convention regarding this in `yapapi` and I imagine it would be non-trivial to enforce it, especially since using double underscores has actual side-effects on the field. I'd leave this as is.